source/common/network/connection_impl.cc

```diff
@@ -438,6 +450,27 @@ void ConnectionImpl::onWriteReady() {
     ASSERT(0 == rc);
     UNREFERENCED_PARAMETER(rc);
 
+#if !defined(__linux__)
+    // When a connect() is performed on a non-blocking socket, it'll return -1/EINPROGRESS
+    // immediately and when select()'ed for writability it'll succeed or fail but when it fails the
+    // reason is hidden. `getsockopt(,,SO_ERROR,,)` solves this issue on some systems but on some
+    // cases such as FreeBSD getsockopt returns 0 even on `EAGAIN`. The following trick issues a
+    // `getpeername()` which will return -1/ENOTCONN if the socket is not connected and read(,,1)
+    // will produce the right errno.
```

Comment:
I believe that connection error will be available in `kq_errno` on OSX & FreeBSD and should be used instead of `getsockopt(SO_ERROR)` on those platforms, so this might not be necessary at all.